src/reducers.ts

```diff
@@ -51,3 +53,9 @@ function moveShape(state: State, action: MoveShapeAction) {
     })
   });
 }
+
+function bringShapeForward(state: State, action: BringShapeForwardAction) {
```

Comment:
I have to admit, that I don't have the best answer to where these functions would go. But, the idea is that your app defines all the functions to transform the model and state, and you reuse these in your Redux reducers. This way you minimize the decoupling of the app to Redux, making it easier to migrate away if and when.

So, one approach is to create an `Application` model that you'd put in your model's folder. In there, you would have a function `bringShapeForward(application: Application): Application`.

```ts
// ./src/models/application.ts

interface Application {
  shapes: Shape[];
}

function bringShapeForward(application: Application, shapeIndex: number): Application {
  const shapes = application.shapes;
  return Object.assign({}, application, {shapes: [...shapes, shapes[shapeIndex]]});
}
```

Then, in your reducer, you'd just use this method:

```ts
import * as models from "./src/models/application.ts";

export function reducer(state: models.Application, action: Action): models.Application {
  if (action.type == "BRING_FORWARD") {
    return models.bringShapeForward(state, (action as BringForwardAction).shapeIndex);
  } else {
    return state;
}
```